src/category_theory/abelian/pseudoelements.lean

```diff
@@ -0,0 +1,411 @@
+
+/-- If two elements are pseudo-equal, then their composition with a morphism is, too. -/
+lemma pseudo_apply_aux {P Q : C} (f : P ⟶ Q) (a b : over P) :
+  a ≈ b → app f a ≈ app f b :=
+λ ⟨R, p, q, ep, eq, comm⟩,
+  ⟨R, p, q, ep, eq, show p ≫ a.hom ≫ f = q ≫ b.hom ≫ f, by rw reassoc_of comm⟩
+
+/-- A morphism `f` induces a function `pseudo_apply f` on pseudoelements. -/
+def pseudo_apply {P Q : C} (f : P ⟶ Q) : P → Q :=
+quotient.map (λ (g : over P), app f g) (pseudo_apply_aux f)
+
+/-- A coercion from morphisms to functions on pseudoelements -/
+def hom_to_fun {P Q : C} : has_coe_to_fun (P ⟶ Q) := ⟨_, pseudo_apply⟩
+
+local attribute [instance] hom_to_fun
+
+lemma pseudo_apply_bar {P Q : C} (f : P ⟶ Q) (a : over P) : f ⟦a⟧ = ⟦a.hom ≫ f⟧ :=
+rfl
+
+/-- Applying a pseudoelement to a composition of morphisms is the same as composing
+    with each morphism. Sadly, this is not a definitional equality, but at least it is
+    true. -/
+theorem comp_apply {P Q R : C} (f : P ⟶ Q) (g : Q ⟶ R) (a : P) : (f ≫ g) a = g (f a) :=
+quotient.induction_on a $ λ x, quotient.sound $ by { unfold app, rw [←category.assoc, over.coe_hom] }
+
+/-- Composition of functions on pseudoelements is composition of morphisms. -/
+theorem comp_comp {P Q R : C} (f : P ⟶ Q) (g : Q ⟶ R) : g ∘ f = f ≫ g :=
+funext $ λ x, (comp_apply _ _ _).symm
+
+section zero
+/-!
+In this section we prove that for every `P` there is an equivalence class that contains
+precisely all the zero morphisms ending in `P` and use this to define *the* zero
+pseudoelement.
+-/
+
+section
+local attribute [instance] has_binary_biproducts.of_has_binary_products
+
+/-- The arrows pseudo-equal to a zero morphism are precisely the zero morphisms -/
+lemma pseudo_zero_aux {P : C} (Q : C) (f : over P) : f ≈ (0 : Q ⟶ P) ↔ f.hom = 0 :=
+⟨λ ⟨R, p, q, ep, eq, comm⟩, by exactI zero_of_epi_comp p (by simp [comm]),
+  λ hf, ⟨biprod f.1 Q, biprod.fst, biprod.snd, by apply_instance, by apply_instance,
+    by rw [hf, over.coe_hom, has_zero_morphisms.comp_zero, has_zero_morphisms.comp_zero]⟩⟩
+
+end
+
+lemma zero_eq_zero' {P Q R : C} :
+  ⟦((0 : Q ⟶ P) : over P)⟧ = ⟦((0 : R ⟶ P) : over P)⟧ :=
+quotient.sound $ (pseudo_zero_aux R _).2 rfl
+
+/-- The zero pseudoelement is the class of a zero morphism -/
+def pseudo_zero {P : C} : P := ⟦(0 : P ⟶ P)⟧
+
+instance {P : C} : has_zero P := ⟨pseudo_zero⟩
+instance {P : C} : inhabited (pseudoelement P) := ⟨0⟩
+
+lemma pseudo_zero_def {P : C} : (0 : pseudoelement P) = ⟦(0 : P ⟶ P)⟧ := rfl
+@[simp] lemma zero_eq_zero {P Q : C} : ⟦((0 : Q ⟶ P) : over P)⟧ = (0 : pseudoelement P) :=
+zero_eq_zero'
+
+/-- The pseudoelement induced by an arrow is zero precisely when that arrow is zero -/
+lemma pseudo_zero_iff {P : C} (a : over P) : (a : P) = 0 ↔ a.hom = 0 :=
+by { rw ←pseudo_zero_aux P a, exact quotient.eq }
+
+end zero
+
+/-- Morphisms map the zero pseudoelement to the zero pseudoelement -/
+theorem apply_zero {P Q : C} (f : P ⟶ Q) : f 0 = 0 :=
```

Comment:
Should these be `simp` lemmas? (I'm not sure about some of the lemmas above. But they might qualify as well.)